### PR TITLE
Feature auth http requests

### LINKED ISSUE #613: support auth HTTP requests in default Comunica
#### Issue type:

- :heavy_plus_sign: Feature request

____
#### Description:
It would be nice if the default Comunica query engine also allowed to query Linked Data Fragments that need authentication, e.g. a secured triplestore needing basic auth.

According to question https://github.com/comunica/comunica/issues/612, it should be doable to add a HTTP actor that can handle authenticated HTTP requests, e.g. by starting from [`actor-http-solid-auth-fetch`](https://github.com/comunica/actor-http-solid-auth-fetch).

### PR DESCRIPTION
**Pull request by FlorianFV**
Closes #613

### CODE REVIEW

File: packages/actor-init-sparql/test/ActorInitSparql-test.ts [hunk 1]
```diff
@@ -91,6 +91,7 @@ describe('ActorInitSparql', () => {
   describe('An ActorInitSparql instance', () => {
     const hypermedia = 'http://example.org/';
     const hypermedia2 = 'hypermedia@http://example.org/';
+    const hypermedia3 = 'hypermedia@http://username:passwd@example.org/';
```
Comment (rubensworks) [inline packages/actor-init-sparql/test/ActorInitSparql-test.ts:94]: Could you also add a test for `http://username:passwd@example.org/`?

File: packages/actor-init-sparql/test/ActorInitSparql-test.ts [hunk 2]
```diff
@@ -475,6 +476,18 @@ describe('ActorInitSparql', () => {
         });
     });
 
+    it('should run with a tagged hypermedia and credentials in url and query file option from argv', () => {
```
Comment (rubensworks) [inline packages/actor-init-sparql/test/ActorInitSparql-test.ts:479]: We want to make sure that the resulting context contains the correct credentials and source. Something like in this test: https://github.com/comunica/comunica/blob/d5edecea04f6b25d4235bea09dd6b3b7a1b8a12e/packages/actor-init-sparql/test/ActorInitSparql-test.ts#L759

File: packages/bus-rdf-resolve-quad-pattern/lib/ActorRdfResolveQuadPattern.ts [hunk 3]
```diff
@@ -108,6 +117,8 @@ IActorRdfResolveQuadPatternOutput> {
 export type IDataSource = string | RDF.Source | {
   type?: string;
   value: string | RDF.Source;
+  username?: string;
```
Comment (rubensworks) [inline packages/bus-rdf-resolve-quad-pattern/lib/ActorRdfResolveQuadPattern.ts:120]: I would not mention `username` and `password` here explicitly. A more extensible way of working would be to use a `sourceContext?: ActionContext` field, which contains additional context entries specifically for this source (explain this as a jsdoc comment as well).

File: packages/bus-rdf-resolve-quad-pattern/lib/ActorRdfResolveQuadPattern.ts [hunk 4]
```diff
@@ -27,6 +27,15 @@ export function getDataSourceType(dataSource: IDataSource): string | undefined {
 export function getDataSourceValue(dataSource: IDataSource): string | RDF.Source {
   return isDataSourceRawType(dataSource) ? dataSource : dataSource.value;
 }
+export function getExtraSourceParams(dataSource: IDataSource): any {
```
Comment (rubensworks) [inline packages/bus-rdf-resolve-quad-pattern/lib/ActorRdfResolveQuadPattern.ts:30]: (make sure to read the next comment first)

I would rename this to `getDataSourceContext`, give it a second param (`originalContext: ActionContext`), and return a new ActionContext that is the merge of the original context and the datasource-specific context. If no datasource-specific context exist, you can just return the original context.

File: packages/actor-http-native/lib/Requester.ts [hunk 5]
```diff
@@ -27,7 +27,7 @@ export default class Requester {
   public createRequest(settings: any): EventEmitter {
     // Parse the request URL
     if (settings.url) {
-      Object.assign(settings, url.parse(settings.url));
+      settings = { ...url.parse(settings.url), ...settings };
```
[the remark this answers is not in the document]
Comment (FlorianFV) [reply packages/actor-http-native/lib/Requester.ts:30]: It was because url.parse contained an auth field. My auth in settings was being overwritten.

File: packages/bus-rdf-resolve-quad-pattern/lib/ActorRdfResolveQuadPattern.ts (hunk 3 above)
Comment (rubensworks) [reply packages/bus-rdf-resolve-quad-pattern/lib/ActorRdfResolveQuadPattern.ts:120]: Calling it just `context` is probably fine as well.

File: packages/actor-http-native/lib/ActorHttpNative.ts [hunk 6]
```diff
@@ -58,6 +59,14 @@ export class ActorHttpNative extends ActorHttp {
       options.withCredentials = true;
     }
 
+    if (action.context && action.context.get(KEY_CONTEXT_SOURCE)) {
```
[the remark this answers is not in the document]
Comment (FlorianFV) [reply packages/actor-http-native/lib/ActorHttpNative.ts:62]: Is it not possible to have different username+passwds for different sources?

[the remark this answers is not in the document]
Comment (rubensworks) [reply packages/actor-http-native/lib/ActorHttpNative.ts:62]: Yes, but this is something the federated actor should already take care of.

The http actors should not be aware of the concept of a source.

File: packages/actor-http-native/test/ActorHttpNative-test.ts [hunk 7]
```diff
@@ -188,6 +189,18 @@ describe('ActorHttpNative', () => {
       });
       expect(results.body).toMatchObject({ withCredentials: true });
     });
+
+    it('should run with authorization', async() => {
+      mockSetup({ statusCode: 404 });
+      const results: any = await actor.run({
+        input: new Request('http://example.com'),
+        context: ActionContext({
+          [KEY_CONTEXT_SOURCE]: { value: 'http://example.com' },
```
Comment (rubensworks) [inline packages/actor-http-native/test/ActorHttpNative-test.ts:198]: This entry be removed now.

File: packages/actor-http-node-fetch/test/ActorHttpNodeFetch-test.ts [hunk 8]
```diff
@@ -82,5 +83,56 @@ describe('ActorHttpNodeFetch', () => {
       });
       expect(spy).toHaveBeenCalledWith({ url: 'https://www.google.com/' }, { credentials: 'include' });
     });
+
+    it('should run with authorization', async() => {
+      const spy = jest.spyOn(global, 'fetch');
+      await actor.run({
+        input: <Request> { url: 'https://www.google.com/' },
+        context: ActionContext({
+          [KEY_CONTEXT_SOURCE]: { value: 'https://www.google.com/' },
```
Comment (rubensworks) [inline packages/actor-http-node-fetch/test/ActorHttpNodeFetch-test.ts:92]: Idem

File: packages/actor-http-node-fetch/test/ActorHttpNodeFetch-test.ts [hunk 9]
```diff
@@ -82,5 +83,56 @@ describe('ActorHttpNodeFetch', () => {
...
+      );
+    });
+
+    it('should run with authorization and init.headers undefined', async() => {
+      const spy = jest.spyOn(global, 'fetch');
+      await actor.run({
+        input: <Request> { url: 'https://www.google.com/' },
+        init: {},
+        context: ActionContext({
+          [KEY_CONTEXT_SOURCE]: { value: 'https://www.google.com/' },
```
Comment (rubensworks) [inline packages/actor-http-node-fetch/test/ActorHttpNodeFetch-test.ts:108]: Idem

File: packages/actor-http-node-fetch/test/ActorHttpNodeFetch-test.ts [hunk 10]
```diff
@@ -82,5 +83,56 @@ describe('ActorHttpNodeFetch', () => {
...
+    });
+
+    it('should run with authorization and already header in init', async() => {
+      const spy = jest.spyOn(global, 'fetch');
+
+      await actor.run({
+        input: <Request> { url: 'https://www.google.com/' },
+        init: { headers: new Headers({ 'Content-Type': 'image/jpeg' }) },
+        context: ActionContext({
+          [KEY_CONTEXT_SOURCE]: { value: 'https://www.google.com/' },
```
Comment (rubensworks) [inline packages/actor-http-node-fetch/test/ActorHttpNodeFetch-test.ts:125]: Idem

File: packages/actor-init-sparql/lib/ActorInitSparql.ts [hunk 11]
```diff
@@ -69,6 +71,28 @@ export class ActorInitSparql extends ActorInitSparqlBrowser {
     return existsSync(`${__dirname}/../test`);
   }
 
+  // Function to convert a URL like 'hypermedia@http://user:passwd@example.com to an IDataSource
```
Comment (rubensworks) [inline packages/actor-init-sparql/lib/ActorInitSparql.ts:74]: Could you transform this comment into actual jsdoc? So `/** ...`, with `@param` and stuff.

File: packages/actor-init-sparql/lib/ActorInitSparql.ts [hunk 12]
```diff
@@ -69,6 +71,28 @@ export class ActorInitSparql extends ActorInitSparqlBrowser {
...
+  public static getSourceObjectFromString(sourceString: string): {[id: string]: any} {
+    const source: {[id: string]: any} = {};
+    const mediaTypeRegex = /^[^:]*@/u;
+    const mediaTypeMatches = mediaTypeRegex.exec(sourceString);
+    if (mediaTypeMatches) {
+      source.type = mediaTypeMatches[0].slice(0, -1);
```
Comment (rubensworks) [inline packages/actor-init-sparql/lib/ActorInitSparql.ts:80]: Slice would not be needed anymore if you use a capturing group in the regex.

File: packages/actor-init-sparql/lib/ActorInitSparql.ts [hunk 13]
```diff
@@ -69,6 +71,28 @@ export class ActorInitSparql extends ActorInitSparqlBrowser {
...
+      sourceString = sourceString.slice(sourceString.split('@')[0].length + 1);
```
Comment (rubensworks) [inline packages/actor-init-sparql/lib/ActorInitSparql.ts:81]: Split is not needed anymore. It is more performance to `slice` starting from the length of `source.type` (+1).

File: packages/actor-init-sparql/lib/ActorInitSparql.ts [hunk 14]
```diff
@@ -69,6 +71,28 @@ export class ActorInitSparql extends ActorInitSparqlBrowser {
...
+    }
+    const authRegex = /\/\/.*:.*@/u;
+    const authMatches = authRegex.exec(sourceString);
+    if (authMatches) {
+      const credentials = authMatches[0].slice(2, -1);
```
Comment (rubensworks) [inline packages/actor-init-sparql/lib/ActorInitSparql.ts:86]: Same here, use a capturing group instead.

File: packages/actor-init-sparql/lib/ActorInitSparql.ts [hunk 15]
```diff
@@ -69,6 +71,28 @@ export class ActorInitSparql extends ActorInitSparqlBrowser {
...
+      source.context = ActionContext({
+        [KEY_CONTEXT_AUTH]: decodeURIComponent(credentials),
+      });
+      sourceString = `${sourceString.split('@')[0].split('//')[0]}//${sourceString.split('@')[1]}`;
```
Comment (rubensworks) [inline packages/actor-init-sparql/lib/ActorInitSparql.ts:90]: Splits here can also go for better performance.

File: packages/actor-init-sparql/test/ActorInitSparql-test.ts [hunk 16]
```diff
@@ -91,6 +92,8 @@ describe('ActorInitSparql', () => {
   describe('An ActorInitSparql instance', () => {
     const hypermedia = 'http://example.org/';
     const hypermedia2 = 'hypermedia@http://example.org/';
+    const hypermedia3 = 'http://username:passwd@example.org/';
+    const hypermedia4 = 'hypermedia@http://username:passwd@example.org/';
```
Comment (rubensworks) [inline packages/actor-init-sparql/test/ActorInitSparql-test.ts:96]: Could you add some additional tests to check for more edge cases? Like empty username, empty password, encoded username, encoded password, ...

Might be easier to test this at `getSourceObjectFromString` directly.

File: packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts [hunk 17]
```diff
@@ -230,6 +231,7 @@ export class FederatedQuadSource implements ILazyQuadSource {
         let pattern: Algebra.Pattern | undefined;
 
         // Prepare the context for this specific source
+        this.contextDefault = getDataSourceContext(source, this.contextDefault);
```
Comment (rubensworks) [inline packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts:234]: We don't want this to be overridden, since `contextDefault` is the context that is common across all sources.
Only in the scope of the current source do we want the source-specific context to be applied.

File: packages/bus-rdf-resolve-quad-pattern/lib/ActorRdfResolveQuadPattern.ts [hunk 18]
```diff
@@ -14,6 +14,11 @@ export const KEY_CONTEXT_SOURCES = '@comunica/bus-rdf-resolve-quad-pattern:sourc
  * @value {IDataSource} A source.
  */
 export const KEY_CONTEXT_SOURCE = '@comunica/bus-rdf-resolve-quad-pattern:source';
+/**
+ * @type {string} Context entry for the authentication for a source.
+ * @value {string} "username:password"-pair.
+ */
+export const KEY_CONTEXT_AUTH = '@comunica/bus-rdf-resolve-quad-pattern:auth';
```
Comment (rubensworks) [inline packages/bus-rdf-resolve-quad-pattern/lib/ActorRdfResolveQuadPattern.ts:21]:
```suggestion
export const KEY_CONTEXT_AUTH = '@comunica/bus-http:auth';
```

Comment (rubensworks) [reply packages/bus-rdf-resolve-quad-pattern/lib/ActorRdfResolveQuadPattern.ts:21]: Could you also add a context key shortcut here?: https://github.com/comunica/comunica/blob/master/packages/actor-init-sparql/components/Actor/Init/Sparql.jsonld#L108-L118

File: packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts (hunk 17 above)
Comment (FlorianFV) [reply packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts:234]: On the next line, the actor also sets the 'source' field (also source-specific) to the same context. That's why I did it this way.
To what context should it be changed then?

Comment (rubensworks) [reply packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts:234]: What you put in the context is correct, you just shouldn't override `this.contextDefault`. Instead, just put it in the local `context` directly.

Comment (FlorianFV) [reply packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts:234]: So the key-values need to be in this.contextDefault, but just not set in a way that requires an override?

Comment (rubensworks) [reply packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts:234]: No, key-values should be in the local `context`.

Comment (FlorianFV) [reply packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts:234]: Nevermind, I got it

File: packages/actor-init-sparql/components/Actor/Init/Sparql.jsonld [hunk 19]
```diff
@@ -115,7 +115,8 @@
             { "contextShortcutKey": "queryTimestamp",         "contextShortcutValue": "@comunica/actor-init-sparql:queryTimestamp"     },
             { "contextShortcutKey": "httpProxyHandler",       "contextShortcutValue": "@comunica/actor-http-proxy:httpProxyHandler"    },
             { "contextShortcutKey": "lenient",                "contextShortcutValue": "@comunica/actor-init-sparql:lenient"            },
-            { "contextShortcutKey": "httpIncludeCredentials", "contextShortcutValue": "@comunica/bus-http:include-credentials"         }
+            { "contextShortcutKey": "httpIncludeCredentials", "contextShortcutValue": "@comunica/bus-http:include-credentials"         },
+            { "contextShortcutKey": "auth",                   "contextShortcutValue": "@comunica/bus-http:auth"                      }
```
Comment (rubensworks) [inline packages/actor-init-sparql/components/Actor/Init/Sparql.jsonld:119]:
```suggestion
            { "contextShortcutKey": "httpAuth",                   "contextShortcutValue": "@comunica/bus-http:auth"                    }
```

File: packages/actor-init-sparql/components/Actor/Init/Sparql.jsonld [hunk 20]
```diff
@@ -115,7 +115,8 @@
             { "contextShortcutKey": "queryTimestamp",         "contextShortcutValue": "@comunica/actor-init-sparql:queryTimestamp"     },
             { "contextShortcutKey": "httpProxyHandler",       "contextShortcutValue": "@comunica/actor-http-proxy:httpProxyHandler"    },
             { "contextShortcutKey": "lenient",                "contextShortcutValue": "@comunica/actor-init-sparql:lenient"            },
-            { "contextShortcutKey": "httpIncludeCredentials", "contextShortcutValue": "@comunica/bus-http:include-credentials"         }
+            { "contextShortcutKey": "httpIncludeCredentials", "contextShortcutValue": "@comunica/bus-http:include-credentials"         },
+            { "contextShortcutKey": "httpAuth",                   "contextShortcutValue": "@comunica/bus-http:auth"                    }
```
Comment (rubensworks) [inline packages/actor-init-sparql/components/Actor/Init/Sparql.jsonld:119]:
```suggestion
            { "contextShortcutKey": "httpAuth",               "contextShortcutValue": "@comunica/bus-http:auth"                        }
```

File: packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts [hunk 21]
```diff
@@ -21,7 +22,7 @@ export class FederatedQuadSource implements ILazyQuadSource {
   IActorRdfResolveQuadPatternOutput>, IActionRdfResolveQuadPattern, IActorTest, IActorRdfResolveQuadPatternOutput>;
 
   protected readonly sources: DataSources;
-  protected readonly contextDefault: ActionContext;
+  protected contextDefault: ActionContext;
```
Comment (rubensworks) [inline packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts:25]:
```suggestion
  protected readonly contextDefault: ActionContext;
```

File: packages/bus-rdf-resolve-quad-pattern/lib/ActorRdfResolveQuadPattern.ts [hunk 22]
```diff
@@ -14,6 +14,11 @@ export const KEY_CONTEXT_SOURCES = '@comunica/bus-rdf-resolve-quad-pattern:sourc
  * @value {IDataSource} A source.
  */
 export const KEY_CONTEXT_SOURCE = '@comunica/bus-rdf-resolve-quad-pattern:source';
+/**
+ * @type {string} Context entry for the authentication for a source.
+ * @value {string} "username:password"-pair.
+ */
+export const KEY_CONTEXT_AUTH = '@comunica/bus-http:auth';
```
Comment (rubensworks) [inline packages/bus-rdf-resolve-quad-pattern/lib/ActorRdfResolveQuadPattern.ts:21]: Could you put this in the http bus instead? There should already be some context keys in its main abstract actor AFAIK.
